### DISCUSSION TITLE
Production build/deploy & Turborepo setup

### DISCUSSION THREAD

**Discussion by Strydom** [vikejs/vike#789] [Discussion]
Hello, I'm trying to setup Vite + vite-plugin-ssr with Express app in a mono-repo managed by Turborepo.
I have the app running in development but when trying to build for production I'm running into issues.

https://github.com/vercel/turbo/tree/main/examples/with-docker

Scripts:
```
"build": "vite build",
"server:prod": "cross-env NODE_ENV=production ts-node src/server/index.ts"
```

First, all the examples I've found while searching are expecting a JS output and do not rely on the ts files after build. I feel like keeping ts around in prod is making things harder. Is there a better way to build and run for prod?

Dockerfile:
```
FROM node:19-alpine AS base
RUN apk add --no-cache libc6-compat
RUN apk update
WORKDIR /app

ENV PNPM_HOME="/root/.local/share/pnpm"
ENV PATH="${PATH}:${PNPM_HOME}"

RUN npm install --global pnpm

FROM base AS builder
WORKDIR /app

RUN pnpm add -g turbo

# Set up working directory
COPY ../../. .
RUN turbo prune --scope=web --docker

# Add lockfile and package.json's of isolated subworkspace
FROM base AS installer
WORKDIR /app

# First install the dependencies (as they change less often)
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
RUN pnpm install

# Build the project
COPY --from=builder /app/out/full/ .
COPY turbo.json turbo.json
RUN pnpm build:web

FROM base AS runner
WORKDIR /app

# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 pnpm
USER pnpm

COPY --from=installer --chown=pnpm:nodejs /app/apps/web ./app/apps/web

WORKDIR /app/apps/web

CMD pnpm run server:prod
```

I've tried to setup a docker file to make building and running the prod app easier, but am running into errors like
```
Error: Cannot find module '/node_modules/.pnpm/cross-env@7.0.3/node_modules/cross-env/src/bin/cross-env.js'
2023-04-17T17:45:31.724484125Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1090:15)
2023-04-17T17:45:31.724486500Z     at Module._load (node:internal/modules/cjs/loader:934:27)
2023-04-17T17:45:31.724488750Z     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
2023-04-17T17:45:31.724491917Z     at node:internal/main/run_main_module:23:47 {
2023-04-17T17:45:31.724494542Z   code: 'MODULE_NOT_FOUND',
2023-04-17T17:45:31.724496875Z   requireStack: []
2023-04-17T17:45:31.724499000Z }
```

Any help will be much appreciated, prod building and deploying documentation seems to be sparse

**Comment by brillout** [comment]
> prod building and deploying documentation seems to be sparse

That's because it's mostly unrelated to VPS as building is done by Vite and deployment depends on your server and deployment strategy.

But let me know if you believe your issue to be related to vite-plugin-ssr (doesn't seem like it) or if you believe additional VPS docs would be helpful.

**Reply by Strydom** [reply]
Circling back... I managed to get it working with docker + a build steps in github workflows.

```
- name: Prune ${{ matrix.app }}
  run: |
    turbo prune --scope=${{ matrix.app }}
    cp Dockerfile-${{ matrix.app }} out/Dockerfile
    cp .dockerignore out/.dockerignore
- name: Install ${{ matrix.app }}
  run: |
    cd out
    pnpm i --frozen-lockfile
- name: Build ${{ matrix.app }}
  run: |
    cd out/apps/web
    pnpm run build
- name: Delete node_modules
  run: |
    cd out
    rm -rf `find . -type d -name node_modules`
- name: Zip ${{ matrix.app }}
  run: |
    cd out
    zip -r ../${{ env.APP_NAME }}.zip .
```

```
FROM node:19-alpine AS base
RUN apk add --no-cache libc6-compat
RUN apk update
WORKDIR /app

ENV PNPM_HOME="/root/.local/share/pnpm"
ENV PATH="${PATH}:${PNPM_HOME}"

RUN npm install --global pnpm

FROM base AS installer
WORKDIR /app

COPY . .
RUN pnpm i --frozen-lockfile

FROM base AS runner
WORKDIR /app

# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 pnpm
USER pnpm

COPY --from=installer --chown=pnpm:nodejs /app .

WORKDIR /app/apps/web

ENV APP_PORT 80
EXPOSE 80
CMD pnpm run server:prod
```

```
"build": "vite build",
"server:prod": "cross-env NODE_ENV=production ts-node src/server/index.ts"
```

This is still not really production optimised, still contains TS files etc.

**Reply by brillout** [reply]
Thanks for circling back. So the issues were unrelated to vite-plugin-ssr, correct? Is there antyhing vite-plugin-ssr('s docs) could have done to improve your DX?

**Reply by Strydom** [reply]
> So the issues were unrelated to vite-plugin-ssr, correct?

The issues were not really issues, rather a lack of docs explaining/showing what actually needs to be done for bundling, so i was just taking a shot in the dark. I ended up just copying the whole code over in docker, and also installing ts-node in production, which is not great.

> Is there antyhing vite-plugin-ssr('s docs) could have done to improve your DX?

I'm still figuring this bit out, maybe once i have a working example with correct bundling I'll know what is missing from the docs. Right now it feels like everything is missing when it comes to a TS deploy 😅 which is why I'm trying to setup a modern example using the latest and greatest bellow ⬇️ 

EDIT: one specific thing that caught me out was the fact my TS server code needed to be built separately. Ended up just not bothering right now and used ts-node in production build...

**Reply by brillout** [reply]
👍 Let's fix it then let's see how the docs can be improved.

**Comment by Strydom** [comment]
I'm now trying to get a serverless deploy working and am again running into issues.
I've created an example [repo](https://github.com/Strydom/vite-production-serverless-build) that I was hoping you might be able to help me with. I feel like I've figured out most of it, but now can't figure out what to do with this error

```
cannot test case insensitive FS, CLIENT_ENTRY does not point to an existing file: .../GitHub/vite-production-serverless-build/.esbuild/.build/dist/client/client.mjs

```

This [example](https://github.com/jamesladd/aws-vite-ssr) doesn't cover the complexities of TS and so wasn't very helpful.

It's also really not clear to me what files we need to keep and what can be pruned from the production bundle.
Vite was complaining about `package.json` not being available in the build path which i found really strange, why does it need that?

**Reply by brillout** [reply]
> It's also really not clear to me what files we need to keep and what can be pruned from the production bundle.

Depends: do you only bundle `dist/` or do you also bundle your server code e.g. `server/some-express-app.ts`?

> Vite was complaining about `package.json` not being available in the build path which i found really strange, why does it need that?

Was it a Vite error or a vite-plugin-ssr one? What's the entire error including error message and error stack?

**Reply by Strydom** [reply]
> Depends: do you only bundle dist/ or do you also bundle your server code e.g. server/some-express-app.ts?

I'm trying to bundle both. Curious, is that not expected for a production app, especially one intended to be run on a Lambda?

> Was it a Vite error or a vite-plugin-ssr one? What's the entire error including error message and error stack?

Seems to come from `vite/dist/node/constants.js` when triggering the lambda function

```
"Error: ENOENT: no such file or directory, open '/Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/package.json'",
"at Object.openSync (node:fs:589:3)",
"at readFileSync (node:fs:457:35)",
"at node_modules/.pnpm/vite@4.3.5_@types+node@18.16.5/node_modules/vite/dist/node/constants.js (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:487506:31)",
"at __init (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:21:56)",
"at node_modules/.pnpm/vite@4.3.5_@types+node@18.16.5/node_modules/vite/dist/node/chunks/dep-934dbc7c.js (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:560714:5)",
"at __init (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:21:56)",
"at node_modules/.pnpm/vite@4.3.5_@types+node@18.16.5/node_modules/vite/dist/node/index.js (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:586280:5)",
"at __init (file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:21:56)",
"at file:///Users/chstrydom/Documents/GitHub/vite-production-serverless-build/.esbuild/.build/src/server/lambda.js:587094:52"
```

**Reply by Strydom** [reply]
I've started a discussion on the `vite` side https://github.com/vitejs/vite/discussions/13212

**Reply by brillout** [reply]
From vite-plugin-ssr's side, if you bundle your server entry file, then it should just work.

If possible, I'd suggest to:
 1. First make the bundle step work (without Docker nor AWS)
 2. Make the Docker part work (without AWS)
 3. Make the AWS part work

I don't know if you're able to apply these steps in isolation though.

But, if you can, then that should help you figure out what tool is causing problems. I suspect most problems to originate from Docker/AWS.

**Reply by Strydom** [reply]
I finally got it working locally with serverless offline 🎉 you can see the changes i needed to make in https://github.com/vitejs/vite/discussions/13212. I'm not sure if you have any ideas/suggestions for better workarounds/improvements?

I see you mentioned https://github.com/hattipjs/hattip somewhere, so might look into that instead of manually doing the switch between serverless and server, unless you think its not ready to be used with `vite` + `vite-plugin-ssr`?

I will keep improving the example repo with the full multi module/app setup I'm going for, using Turbo repo. Feel free to reference it as an example setup in the docs.

**Reply by brillout** [reply]
Congrats 🎉.

Most issues seem to be interop issues between AWS Lambda and the JS ecosystem. So I don't think there is much room for improvements on vite-plugin-ssr's side (beyond the docs as we discussed)?

Yes HatTip works, see the boilerplate generated by https://magne4000.github.io/batijs.github.io/. Keep me updated how it goes.

>  Feel free to reference it as an example 

👌 Will do.

**Reply by brillout** [reply]
> 👌 Will do.

Done https://vite-plugin-ssr.com/aws-lambda.

> I will keep improving the example repo

👍 Keep me updated. Also let me know what kind of documentation would have helped you.
